File: go/examples/basic.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/RevCBH/slim-protocol-core/go/slim"
)

func main() {
	fmt.Println("=== SLIM Protocol - Go Implementation ===\n")

	// Example 1: Simple object
	fmt.Println("Example 1: Simple object")
	data := map[string]interface{}{
		"name":   "Mario",
		"age":    float64(30),
		"active": true,
	}
	jsonBytes, _ := json.Marshal(data)
	slimStr, _ := slim.Encode(data, slim.DefaultEncodeOptions())
	fmt.Printf("JSON: %s\n", string(jsonBytes))
	fmt.Printf("SLIM: %s\n\n", slimStr)

	// Example 2: Array of objects (table format)
	fmt.Println("Example 2: Array of objects (table format)")
	users := []interface{}{
		map[string]interface{}{"id": float64(1), "name": "Mario", "active": true},
		map[string]interface{}{"id": float64(2), "name": "Luigi", "active": false},
		map[string]interface{}{"id": float64(3), "name": "Peach", "active": true},
	}
	jsonBytes, _ = json.Marshal(users)
	slimStr, _ = slim.Encode(users, slim.DefaultEncodeOptions())
	fmt.Printf("JSON: %s\n", string(jsonBytes))
	fmt.Printf("SLIM:\n%s\n\n", slimStr)

	// Example 3: Decode back
	fmt.Println("Example 3: Decode SLIM back to data")
	slimInput := "|2|id#,name$|\n1,Mario\n2,Luigi"
	decoded, _ := slim.Decode(slimInput, slim.DefaultDecodeOptions())
	decodedJSON, _ := json.MarshalIndent(decoded, "", "  ")
	fmt.Printf("SLIM: %s\n", slimInput)
	fmt.Printf("Decoded:\n%s\n\n", string(decodedJSON))

	// Example 4: Token savings
	fmt.Println("Example 4: Token savings comparison")
	data2 := []interface{}{
		map[string]interface{}{"id": float64(1), "score": 95.5, "passed": true},
		map[string]interface{}{"id": float64(2), "score": 87.2, "passed": true},
		map[string]interface{}{"id": float64(3), "score": 62.8, "passed": false},
	}
	jsonStr, _ := json.Marshal(data2)
	slimStr2, _ := slim.Encode(data2, slim.DefaultEncodeOptions())
	fmt.Printf("JSON length: %d chars\n", len(jsonStr))
	fmt.Printf("SLIM length: %d chars\n", len(slimStr2))
	fmt.Printf("Savings:     %.1f%%\n",
		float64(len(jsonStr)-len(slimStr2))/float64(len(jsonStr))*100)

	// Example 5: Schema inference
	fmt.Println("\nExample 5: Schema inference")
	schema := slim.InferSchema(users)
	fmt.Printf("Inferred schema: %s\n", schema)

	// Validate data
	result := slim.ValidateSchema(users, schema)
	if result.Valid {
		fmt.Println("✓ Data is valid according to schema")
	} else {
		fmt.Println("✗ Validation errors:")
		for _, err := range result.Errors {
			fmt.Printf("  - %s: %s\n", err.Path, err.Message)
		}
	}
}

File: rust/examples/basic.rs
//! Basic SLIM encoding/decoding example

use slim_protocol::{encode, decode};
use serde_json::json;

fn main() {
    println!("=== SLIM Protocol - Basic Example ===\n");

    // Example 1: Simple object
    println!("Example 1: Simple object");
    let data = json!({"name": "Mario", "age": 30, "active": true});
    let slim = encode(&data, Default::default()).unwrap();
    println!("Original JSON: {}", serde_json::to_string(&data).unwrap());
    println!("SLIM format:   {}", slim);
    println!();

    // Example 2: Array of objects (table format)
    println!("Example 2: Array of objects (table format)");
    let users = json!([
        {"id": 1, "name": "Mario", "active": true},
        {"id": 2, "name": "Luigi", "active": false},
        {"id": 3, "name": "Peach", "active": true}
    ]);
    let slim = encode(&users, Default::default()).unwrap();
    println!("Original JSON: {}", serde_json::to_string(&users).unwrap());
    println!("SLIM format:\n{}", slim);
    println!();

    // Example 3: Decode back
    println!("Example 3: Decode SLIM back to JSON");
    let slim_str = "|2|id#,name$|\n1,Mario\n2,Luigi";
    let decoded = decode(slim_str, Default::default()).unwrap();
    println!("SLIM:    {}", slim_str.replace('\n', "\\n"));
    println!("Decoded: {}", serde_json::to_string_pretty(&decoded).unwrap());
    println!();

    // Example 4: Token savings
    println!("Example 4: Token savings comparison");
    let data = json!([
        {"id": 1, "score": 95.5, "passed": true},
        {"id": 2, "score": 87.2, "passed": true},
        {"id": 3, "score": 62.8, "passed": false}
    ]);
    let json_str = serde_json::to_string(&data).unwrap();
    let slim_str = encode(&data, Default::default()).unwrap();

    println!("JSON length: {} chars", json_str.len());
    println!("SLIM length: {} chars", slim_str.len());
    println!("Savings:     {:.1}%",
        ((json_str.len() - slim_str.len()) as f64 / json_str.len() as f64) * 100.0);
}
